exp/operator/api/v1alpha4/interfaces.go

```diff
@@ -0,0 +1,39 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha4
+
+import (
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// +k8s:deepcopy-gen=false
+type Provider interface {
+	client.Object
+	conditions.Setter
+	GetSpec() ProviderSpec
+	SetSpec(in ProviderSpec)
+	GetStatus() ProviderStatus
+	SetStatus(in ProviderStatus)
+}
+
+// +k8s:deepcopy-gen=false
+type ProviderList interface {
+	client.ObjectList
+
+	GetItems() []Provider
+}
```

Comment:
I agree we should get those type, however I think we should keep the API surface as clean as possible and avoid to surface controller implementation details at this level.

What about moving those types under controllers / another package?
This will require also to have wrapper types / factory methods instead of implementing the interface methods in the API (more down in the review about this idea)